Stop using deprecated UTType API

brentfulgham opened the pull request:
#### f47f4151b99208bcf4fe9db090f39f0effedc9b4
<pre>
Stop using deprecated UTType API
<a href="https://bugs.webkit.org/show_bug.cgi?id=256867">https://bugs.webkit.org/show_bug.cgi?id=256867</a>
&lt;<a href="https://rdar.apple.com/problem/109431567">rdar://problem/109431567</a>&gt;

Reviewed by Aditya Keerthi.

Update WebKit to use the modern UTType API, rather than the deprecated Carbon-era C APIs.

Also remove WebCore::UTIFromTag and related Internals methods that were added long ago to
support tests for a feature (and tests) that were removed a few years ago.

* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/PAL/pal/system/mac/PopupMenu.mm:
(PAL::popUpMenu):
* Source/WebCore/PAL/pal/system/mac/WebPanel.mm:
(-[WebPanel init]):
* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/editing/cocoa/WebContentReaderCocoa.mm:
(WebCore::mimeTypeFromContentType):
(WebCore::attachmentForFilePath):
* Source/WebCore/editing/mac/EditorMac.mm:
(WebCore::Editor::dataSelectionForPasteboard):
* Source/WebCore/fileapi/FileCocoa.mm:
(WebCore::File::shouldReplaceFile):
* Source/WebCore/page/FrameInlines.h: Copied from Source/WebCore/PAL/pal/system/mac/WebPanel.mm.
(WebCore::Frame::page const):
(WebCore::Frame::ownerElement const):
* Source/WebCore/page/mac/DragControllerMac.mm:
(WebCore::DragController::updateSupportedTypeIdentifiersForDragHandlingMethod const):
* Source/WebCore/platform/cocoa/DragDataCocoa.mm:
(WebCore::rtfPasteboardType):
(WebCore::rtfdPasteboardType):
(WebCore::stringPasteboardType):
(WebCore::urlPasteboardType):
(WebCore::htmlPasteboardType):
(WebCore::pdfPasteboardType):
(WebCore::tiffPasteboardType):
(WebCore::DragData::containsCompatibleContent const):
* Source/WebCore/platform/cocoa/PasteboardCocoa.mm:
(WebCore::cocoaTypeToImageType):
(WebCore::Pasteboard::fileContentState):
* Source/WebCore/platform/cocoa/PlatformPasteboardCocoa.mm:
(WebCore::PlatformPasteboard::urlStringSuitableForLoading):
* Source/WebCore/platform/graphics/cg/UTIRegistry.mm: Renamed from Source/WebCore/platform/graphics/cg/UTIRegistry.cpp.
(WebCore::preferredExtensionForImageType):
* Source/WebCore/platform/ios/LegacyTileGrid.mm:
(WebCore::LegacyTileGrid::dropDistantTiles):
* Source/WebCore/platform/ios/PasteboardIOS.mm:
(WebCore::Pasteboard::writePlainText):
(WebCore::Pasteboard::read):
(WebCore::supportedImageTypes):
(WebCore::isTypeAllowedByReadingPolicy):
(WebCore::Pasteboard::readPasteboardWebContentDataForType):
(WebCore::Pasteboard::supportedWebContentPasteboardTypes):
(WebCore::Pasteboard::supportedFileUploadPasteboardTypes):
(WebCore::Pasteboard::readPlatformValuesAsStrings):
(WebCore::Pasteboard::addHTMLClipboardTypesForCocoaType):
* Source/WebCore/platform/ios/PlatformPasteboardIOS.mm:
* Source/WebCore/platform/network/mac/UTIUtilities.h:
* Source/WebCore/platform/network/mac/UTIUtilities.mm:
(WebCore::MIMETypeFromUTI):
(WebCore::UTIFromUnknownMIMEType):
(WebCore::UTIFromMIMETypeCachePolicy::createValueForKey):
(WebCore::isDeclaredUTI):
(WebCore::UTIFromTag): Deleted.
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::getUTIFromTag): Deleted.
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebCore/testing/Internals.mm:
(WebCore::Internals::getUTIFromTag): Deleted.

Canonical link: <a href="https://commits.webkit.org/280468@main">https://commits.webkit.org/280468@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a602497bb95423008850c3496fec3e6f6400856d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56720 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36048 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9194 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60335 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7163 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/58847 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43671 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7353 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/45953 "Found 1 new test failure: media/video-src-blob-replay.html (failure)") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5028 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/58750 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/33886 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/48957 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/26810 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30666 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/6288 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6166 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/52612 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6559 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62018 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/633 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/6661 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53215 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/636 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49012 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53234 "Found 1 new API test failure: /WebKitGTK/TestWebKitWebContext:/webkit/WebKitWebContext/proxy (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/544 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8435 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/31877 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/32962 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34047 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32708 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/editing/cocoa/WebContentReaderCocoa.mm:
```diff
@@ -722,15 +722,13 @@ static String typeForAttachmentElement(const String& contentType)
-ALLOW_DEPRECATED_DECLARATIONS_BEGIN
         if (isDirectory)
-            contentType = kUTTypeDirectory;
+            contentType = String(UTTypeDirectory.identifier);
```
No need for `String()` here.

brentfulgham replied:
Thank you for taking the time to review this!  I'm iterating on the patch as I try to work through some adoption issues with the new API. I'll definitely incorporate these changes (and appreciate them!), but I do intend to do a bit more work. I hope I don't want to waste your time. I will do a better job of marking future PR's as DRAFT to help clarify this.

brentfulgham replied:
(By new API I mean "the not-deprecated API" :-) )

pxlcoder replied:
No worries!

pxlcoder commented on Source/WebCore/editing/cocoa/WebContentReaderCocoa.mm:
```diff
@@ -722,15 +722,13 @@ static String typeForAttachmentElement(const String& contentType)
             contentType = File::contentTypeForFile(path);
             if (contentType.isEmpty())
-                contentType = kUTTypeData;
+                contentType = String(UTTypeData.identifier);
```
No need for `String()` here.

pxlcoder commented on Source/WebCore/page/FrameInlines.h:
```diff
@@ -0,0 +1,43 @@
+#include "Frame.h"
+#include "Page.h"
+
+namespace WebCore {
```
I don't think this change belongs in this patch?

brentfulgham replied:
You are right! I'll remove this.

pxlcoder commented on Source/WebCore/PAL/pal/system/mac/PopupMenu.mm:
```diff
@@ -44,9 +44,7 @@ - (void)popUpMenu:(NSMenu*)menu atLocation:(NSPoint)location width:(CGFloat)widt
-ALLOW_DEPRECATED_DECLARATIONS_BEGIN
-    if (controlSize != NSMiniControlSize) {
-ALLOW_DEPRECATED_DECLARATIONS_END
+    if (controlSize != NSControlSizeMini) {
```
Not really related to `UTType`, but a small enough change that this is fine.

pxlcoder commented on Source/WebCore/PAL/pal/system/mac/WebPanel.mm:
```diff
@@ -35,7 +35,7 @@ @implementation WebPanel
 {
 ALLOW_DEPRECATED_DECLARATIONS_BEGIN
-    static NSUInteger styleMask = NSMiniaturizableWindowMask | NSClosableWindowMask | NSResizableWindowMask  | NSTitledWindowMask | NSSmallWindowMask | NSSideUtilityWindowMask | NSUtilityWindowMask;
+    static NSUInteger styleMask = NSWindowStyleMaskMiniaturizable | NSWindowStyleMaskClosable | NSWindowStyleMaskResizable | NSWindowStyleMaskTitled | NSSmallWindowMask | NSSideUtilityWindowMask | NSWindowStyleMaskUtilityWindow;
```
Not really related to `UTType`, but a small enough change that this is fine.

pxlcoder commented on Source/WebCore/platform/cocoa/PlatformPasteboardCocoa.mm:
```diff
@@ -74,10 +75,8 @@
-    String stringPasteboardType = kUTTypeText;
-ALLOW_DEPRECATED_DECLARATIONS_END
+    String urlPasteboardType = String(UTTypeURL.identifier);
+    String stringPasteboardType = String(UTTypeText.identifier);
```
No need for `String()` here.

pxlcoder commented on Source/WebCore/platform/ios/PasteboardIOS.mm:
```diff
@@ -138,9 +138,7 @@ static int64_t changeCountForPasteboard(const String& pasteboardName = { }, cons
-ALLOW_DEPRECATED_DECLARATIONS_BEGIN
-    platformStrategies()->pasteboardStrategy()->writeToPasteboard(kUTTypeText, text, m_pasteboardName, context());
-ALLOW_DEPRECATED_DECLARATIONS_END
+    platformStrategies()->pasteboardStrategy()->writeToPasteboard(String(UTTypeText.identifier), text, m_pasteboardName, context());
```
No need for `String()` here.

pxlcoder commented on Source/WebCore/platform/ios/PasteboardIOS.mm:
```diff
@@ -174,9 +172,7 @@ static int64_t changeCountForPasteboard(const String& pasteboardName = { }, cons
-ALLOW_DEPRECATED_DECLARATIONS_BEGIN
-        text.text = strategy.readStringFromPasteboard(itemIndexToQuery, kUTTypeURL, m_pasteboardName, context());
-ALLOW_DEPRECATED_DECLARATIONS_END
+        text.text = strategy.readStringFromPasteboard(itemIndexToQuery, String(UTTypeURL.identifier), m_pasteboardName, context());
```
No need for `String()` here.

pxlcoder replied:
No need for `String`.

pxlcoder commented on Source/WebCore/platform/ios/PasteboardIOS.mm:
```diff
@@ -187,52 +183,45 @@ static int64_t changeCountForPasteboard(const String& pasteboardName = { }, cons
     // should consider removing support for reading plain text from "public.text".
-ALLOW_DEPRECATED_DECLARATIONS_BEGIN
-    text.text = strategy.readStringFromPasteboard(itemIndexToQuery, kUTTypePlainText, m_pasteboardName, context());
+    text.text = strategy.readStringFromPasteboard(itemIndexToQuery, String(UTTypePlainText.identifier), m_pasteboardName, context());
```
No need for `String()` here.

pxlcoder replied:
No need for `String`.

brentfulgham replied:
Sorry  I missed that -- thank you for catching it. I'll do a double-check before landing.

pxlcoder commented on Source/WebCore/platform/ios/PasteboardIOS.mm:
```diff
@@ -187,52 +183,45 @@ static int64_t changeCountForPasteboard(const String& pasteboardName = { }, cons
     if (text.text.isEmpty())
-        text.text = strategy.readStringFromPasteboard(itemIndexToQuery, kUTTypeText, m_pasteboardName, context());
-ALLOW_DEPRECATED_DECLARATIONS_END
+        text.text = strategy.readStringFromPasteboard(itemIndexToQuery, String(UTTypeText.identifier), m_pasteboardName, context());
```
No need for `String()` here.

pxlcoder replied:
No need for `String`.

pxlcoder commented on Source/WebCore/platform/graphics/cg/UTIRegistry.mm:
```diff
@@ -132,9 +133,14 @@ String MIMETypeForImageType(const String& uti)
+
+    auto *type = [UTType typeWithIdentifier:uti];
+    String extension = type.preferredFilenameExtension;
+    RELEASE_ASSERT(oldExtension == extension);
```
Do we really want to `RELEASE_ASSERT` here?

brentfulgham replied:
I was doing this for local testing to convince myself that the two API's give the same result. Since all tests passed with this, I can remove it now.

pxlcoder commented on Source/WebCore/platform/ios/PasteboardIOS.mm:
```diff
@@ -188,20 +184,16 @@ static int64_t changeCountForPasteboard(const String& pasteboardName = { }, cons
-ALLOW_DEPRECATED_DECLARATIONS_BEGIN
-    return @[(__bridge NSString *)kUTTypePNG, (__bridge NSString *)kUTTypeTIFF, (__bridge NSString *)kUTTypeJPEG, (__bridge NSString *)kUTTypeGIF];
-ALLOW_DEPRECATED_DECLARATIONS_END
+    return @[UTTypePNG.identifier, UTTypeTIFF.identifier, UTTypeJPEG.identifier, UTTypeGIF.identifier];
```
Nit:
```suggestion
    return @[ UTTypePNG.identifier, UTTypeTIFF.identifier, UTTypeJPEG.identifier, UTTypeGIF.identifier ];
```

pxlcoder commented on Source/WebCore/platform/network/mac/UTIUtilities.mm:
```diff
@@ -47,9 +48,8 @@
-ALLOW_DEPRECATED_DECLARATIONS_BEGIN
-    return adoptCF(UTTypeCopyPreferredTagWithClass(uti.createCFString().get(), kUTTagClassMIMEType)).get();
-ALLOW_DEPRECATED_DECLARATIONS_END
+    auto *type = [UTType typeWithIdentifier:uti];
```
```suggestion
    RetainPtr type = [UTType typeWithIdentifier:uti];
```

(since you're changing this line anyway, and we have explicit smart pointer guidelines now)

pxlcoder commented on Source/WebCore/platform/network/mac/UTIUtilities.mm:
```diff
@@ -108,28 +108,26 @@ static CFStringRef UTIFromUnknownMIMEType(StringView mimeType)
-ALLOW_DEPRECATED_DECLARATIONS_END
-        if (type)
-            return type;
+        if (auto *type = [UTType typeWithMIMEType:mimeType])
```
Same comment regarding smart pointer usage.

pxlcoder commented on Source/WebCore/platform/network/mac/UTIUtilities.mm:
```diff
@@ -141,17 +139,8 @@ String UTIFromMIMEType(const String& mimeType)
-    auto u = adoptCF(UTTypeCreatePreferredIdentifierForTag(tagClass.createCFString().get(), tag.createCFString().get(), conformingToUTI.createCFString().get()));
-ALLOW_DEPRECATED_DECLARATIONS_END
-    return u.get();
+    auto *type = [UTType typeWithIdentifier:UTI];
```
Same comment regarding smart pointer usage.

brentfulgham committed:
Stop using deprecated UTType API
https://bugs.webkit.org/show_bug.cgi?id=256867
<rdar://problem/109431567>

Reviewed by Aditya Keerthi.

Update WebKit to use the modern UTType API, rather than the deprecated Carbon-era C APIs.

Also remove WebCore::UTIFromTag and related Internals methods that were added long ago to
support tests for a feature (and tests) that were removed a few years ago.

* Source/WebCore/Configurations/WebCore.xcconfig:
* Source/WebCore/PAL/pal/system/mac/PopupMenu.mm:
(PAL::popUpMenu):
* Source/WebCore/PAL/pal/system/mac/WebPanel.mm:
(-[WebPanel init]):
* Source/WebCore/PlatformMac.cmake:
* Source/WebCore/SourcesCocoa.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/editing/cocoa/WebContentReaderCocoa.mm:
(WebCore::mimeTypeFromContentType):
(WebCore::attachmentForFilePath):
* Source/WebCore/editing/mac/EditorMac.mm:
(WebCore::Editor::dataSelectionForPasteboard):
* Source/WebCore/fileapi/FileCocoa.mm:
(WebCore::File::shouldReplaceFile):
* Source/WebCore/page/FrameInlines.h: Copied from Source/WebCore/PAL/pal/system/mac/WebPanel.mm.
(WebCore::Frame::page const):
(WebCore::Frame::ownerElement const):
* Source/WebCore/page/mac/DragControllerMac.mm:
(WebCore::DragController::updateSupportedTypeIdentifiersForDragHandlingMethod const):
* Source/WebCore/platform/cocoa/DragDataCocoa.mm:
(WebCore::rtfPasteboardType):
(WebCore::rtfdPasteboardType):
(WebCore::stringPasteboardType):
(WebCore::urlPasteboardType):
(WebCore::htmlPasteboardType):
(WebCore::pdfPasteboardType):
(WebCore::tiffPasteboardType):
(WebCore::DragData::containsCompatibleContent const):
* Source/WebCore/platform/cocoa/PasteboardCocoa.mm:
(WebCore::cocoaTypeToImageType):
(WebCore::Pasteboard::fileContentState):
* Source/WebCore/platform/cocoa/PlatformPasteboardCocoa.mm:
(WebCore::PlatformPasteboard::urlStringSuitableForLoading):
* Source/WebCore/platform/graphics/cg/UTIRegistry.mm: Renamed from Source/WebCore/platform/graphics/cg/UTIRegistry.cpp.
(WebCore::preferredExtensionForImageType):
* Source/WebCore/platform/ios/LegacyTileGrid.mm:
(WebCore::LegacyTileGrid::dropDistantTiles):
* Source/WebCore/platform/ios/PasteboardIOS.mm:
(WebCore::Pasteboard::writePlainText):
(WebCore::Pasteboard::read):
(WebCore::supportedImageTypes):
(WebCore::isTypeAllowedByReadingPolicy):
(WebCore::Pasteboard::readPasteboardWebContentDataForType):
(WebCore::Pasteboard::supportedWebContentPasteboardTypes):
(WebCore::Pasteboard::supportedFileUploadPasteboardTypes):
(WebCore::Pasteboard::readPlatformValuesAsStrings):
(WebCore::Pasteboard::addHTMLClipboardTypesForCocoaType):
* Source/WebCore/platform/ios/PlatformPasteboardIOS.mm:
* Source/WebCore/platform/network/mac/UTIUtilities.h:
* Source/WebCore/platform/network/mac/UTIUtilities.mm:
(WebCore::MIMETypeFromUTI):
(WebCore::UTIFromUnknownMIMEType):
(WebCore::UTIFromMIMETypeCachePolicy::createValueForKey):
(WebCore::isDeclaredUTI):
(WebCore::UTIFromTag): Deleted.
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::getUTIFromTag): Deleted.
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebCore/testing/Internals.mm:
(WebCore::Internals::getUTIFromTag): Deleted.

Canonical link: https://commits.webkit.org/280468@main
